### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by 78fatemanazari**
- Create a class `Classroom` with the following:
  - `@label` instance variable, should be initialized in the constructor.
  - Setter and getter for `@label` (remember about `attr_accessor`).
- Create the `has-many`/`belongs-to` relationship between `Classroom` and `Student`. The following should be implemented:
  - Create the `has-many` side (a classroom has many students).
  - Create the `belongs-to` side (a student belongs to a classroom).
  - Make sure that when adding a student to a classroom it also sets the classroom for the student.
  - Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
- Create a class `Book` with the following:
  - `@title` and `@author` instance variables, should be initialized in the constructor.
  - Setters and getters for instance variables (remember about `attr_accessor`).
- Create a class `Rental` with the following:
  - `@date` instance variable, should be initialized in the constructor.
  - Setter and getter for `@date` (remember about `attr_accessor`).
- Create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. The following should be implemented:
  - Create the `has-many` side of `Book` and `Rental` (a book has many rentals).
  - Create the `belongs-to` side of `Rental` and `Book` (a rental belongs to a book).
  - Create the `has-many` side of `Person` and `Rental` (a person has many rentals).
  - Create the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person).
  - Modify the constructor of `Rental` so `Book` and `Person` are set in it.

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
 class Student < Person
   attr_accessor :classroom
 
   def initialize(age, classroom, name: 'Unknown', parent_permission: true)
     super(age, name, parent_permission)
     @classroom = classroom
+    classroom.add_student(self)
   end
 
   def play_hooky
```
Comment (Strangeal) [inline student.rb]: - [ ] Great progress in your implementation! 🚀 However, I noticed a requirement that seems to be missing in your `Student` class. As per the project specifications, you are expected to define a `method` that serves as the `belongs-to` side (a student belongs to a classroom)`, meaning that a student belongs to a classroom. To ensure the completeness of the association between students and classrooms, it's essential to add this method to your code.

Happy coding! 😊👍

File: classroom.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (Strangeal) [inline classroom.rb]: - [ ] Great job so far in your implementation! 🚀 However, it seems that your `add_rental` method might not be working as expected. The `add_rental` method should take two parameters, `date` and `person`. Additionally, the book itself should also be passed as a parameter to create the new rental, achieving the 'has-many' association. 😊

To achieve this, consider implementing your method like this:

```ruby
def rent_book(person, date)
  Rental.new(date, person, self)
end
```

This revised approach will ensure that the rental is associated with the book correctly. 
Happy coding! 😊🚀👨‍💻

File: person.rb [hunk 3]
```diff
@@ -34,4 +36,8 @@ def can_use_services?
   def correct_name
     @name
   end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
 end
```
Comment (Strangeal) [inline person.rb]: - [ ] Great job so far in your implementation! 🚀 However, it seems that your `add_book_rental` method might not be working as expected. The `add_book_rental` method should take two parameters, `date` and `book`. Additionally, the person itself should also be passed as a parameter to create the new rental, achieving the 'has-many' association. 😊

To achieve this, consider implementing your method like this:

```ruby
  def add_book_rental(book, date)
     Rental.new(date, book, self)
  end
```

This revised approach will ensure that the rental is associated with the book correctly. 

Happy coding! 😊🚀👨‍💻

**Review by Strangeal — CHANGES_REQUESTED** [review]
Hi @78fatemanazari, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implement the features ✅ 
- Professional `README` template ✅ 
- Correct `GitHub` workflow ✅ 
- Meaningful **Github** commits ✅ 
- Correct `.github/workflows` for linters ✅ 
- Followed *best-practices* ✅ 
- Well structured project ✅ 

## Required Changes ♻️
- [ ] Great work on your code! 🚀 I can see you've put in some solid effort. 🙌 However, I have a friendly suggestion to make your code even better.
   - [ ] I'd recommend creating each class in its separate file. 📂 This simple practice can work wonders for your project in terms of organization and maintainability. It makes the codebase easier to navigate and update.

Happy coding! 🌟🚀👨‍💻

---

This comment maintains a positive and friendly tone while providing constructive feedback. It also offers help and encouragement, which can be motivating for the developer.
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
+
```
Comment (BenMukebo) [inline book.rb]: - [OPTIONAL] Great work so far 👍🏻 However the *Getters and Setters for @rentals should be present as well to enable the read and write options. You may update your code like this:

```ruby
  attr_accessor :title, :author, :rentals
```

File: person.rb [hunk 5]
```diff
@@ -1,16 +1,18 @@
 require_relative 'nameable'
 require_relative 'decorator'
+require_relative 'classroom'
 
 class Person < Nameable
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rentals
```
Comment (BenMukebo) [inline person.rb:8]: => The same here 👇🏻 

> - [OPTIONAL] Great work so far 👍🏻 However the *Getters and Setters for @rentals should be present as well to enable the read and write options. You may update your code like this:
> 
> ```ruby
>   attr_accessor :title, :author, :rentals
> ```

**Review by BenMukebo — APPROVED** [review]
<h1>Status: Approve  :champagne:  :+1: :confetti_ball: 

### **Hi @78fatemanazari**, 

### **Congratulations! 🎉 Project Approved ✅**

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 


### To Highlight :1st_place_medal: 
- Your Pull Request has a proper title and description. ✔️  
- No Linter errors :heavy_check_mark: 
- Your README looks awesome, keep rocking :rocket: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
heers :clinking_glasses:  and Happy coding!!! :dancers:💻  

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**